engine/pkg/election/elector.go

```diff
@@ -0,0 +1,384 @@
+					e.cancelCallback("leader resigned")
+				}
+			} else {
+				req.errCh <- nil
+				e.resignUntil = time.Now().Add(req.duration)
+			}
+		case <-time.After(e.config.RenewInterval):
+		}
+	}
+}
+
+func (e *Elector) renew(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, e.config.RenewDeadline)
+	defer cancel()
+
+	for {
+		err := e.tryRenew(ctx)
+		if err == nil {
+			return nil
+		}
+		randDelay := time.Duration(rand.Int63n(int64(e.config.RenewInterval)))
+		log.Info("renew lease failed, retry after random delay",
+			zap.Duration("delay", randDelay), zap.Error(err))
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(randDelay):
+		}
+	}
+}
+
+func (e *Elector) tryRenew(ctx context.Context) (err error) {
+	start := time.Now()
+	defer func() {
+		log.Debug("tryRenew", zap.Duration("took", time.Since(start)), zap.Error(err))
+	}()
+
+	s := e.config.Storage
+
+	record, err := s.Get(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.setObservedRecord(record)
+
+	var activeMembers []*Member
+	for _, m := range record.Members {
+		if e.isLeaseExpired(m.ID) {
+			if m.ID == record.LeaderID {
+				record.LeaderID = ""
+				log.Info(
+					"leader lease expired",
+					zap.String("leader-id", m.ID),
+					zap.String("leader-name", m.Name),
+					zap.String("leader-address", m.Address),
+				)
+			} else {
+				log.Info(
+					"member lease expired",
+					zap.String("member-id", m.ID),
+					zap.String("member-name", m.Name),
+					zap.String("member-address", m.Address),
+				)
+			}
+		} else {
+			activeMembers = append(activeMembers, m)
+		}
+	}
+	record.Members = activeMembers
+
+	// Add self to the record if not exists.
+	if m, ok := record.FindMember(e.config.ID); !ok {
+		record.Members = append(record.Members, &Member{
+			ID:            e.config.ID,
+			Name:          e.config.Name,
+			Address:       e.config.Address,
+			LeaseDuration: e.config.LeaseDuration,
+			RenewTime:     time.Now(),
+		})
+	} else {
+		m.RenewTime = time.Now()
+	}
+
+	if time.Now().Before(e.resignUntil) {
+		if record.LeaderID == e.config.ID {
+			record.LeaderID = ""
+			log.Info("try to resign leadership")
+		}
+	} else if record.LeaderID == "" {
+		// Elect a new leader if no leader exists.
+		record.LeaderID = e.config.ID
+		log.Info(
+			"try to elect self as leader",
+			zap.String("id", e.config.ID),
+			zap.String("name", e.config.Name),
+			zap.String("address", e.config.Address),
+		)
+	}
+
+	if err := s.Update(ctx, record); err != nil {
+		return errors.Trace(err)
+	}
+	e.setObservedRecord(record)
+	return nil
+}
+
+func (e *Elector) ensureCallbackIsRunning(ctx context.Context) {
+	if !e.callbackIsRunning.Load() {
+		leaderCallback := e.config.LeaderCallback
+		leaderCtx, leaderCancel := context.WithCancel(ctx)
+		e.callbackWg.Add(1)
+		e.callbackIsRunning.Store(true)
+		go func() {
+			defer func() {
+				e.callbackIsRunning.Store(false)
+				e.callbackWg.Done()
+				leaderCancel()
```

Comment:
Can we remove `leaderCancel()` here, since it is always triggered in `leaderCancel`.

Reply:
When the callback is unexpectedly exited, the elector will not cancel the `leaderCtx`.